src/thriveopedia/pages/ThriveopediaEvolutionaryTreePage.cs

```diff
@@ -0,0 +1,158 @@
+    public NodePath SpeciesPreviewPath = null!;
+
+    [Export]
+    public NodePath HexPreviewPath = null!;
+
+    private readonly List<Dictionary<uint, Species>> speciesHistoryList = new();
+    private VBoxContainer disabledInFreebuild = null!;
+    private EvolutionaryTree evolutionaryTree = null!;
+    private CustomRichTextLabel speciesDetailsLabel = null!;
+    private SpeciesPreview speciesPreview = null!;
+    private CellHexesPreview hexesPreview = null!;
+
+    public override string PageName => "EvolutionaryTree";
+    public override string TranslatedPageName => TranslationServer.Translate("EVOLUTIONARY_TREE_PAGE");
+
+    public override void _Ready()
+    {
+        base._Ready();
+
+        disabledInFreebuild = GetNode<VBoxContainer>(DisabledInFreebuildPath);
+        evolutionaryTree = GetNode<EvolutionaryTree>(EvolutionaryTreePath);
+        speciesDetailsLabel = GetNode<CustomRichTextLabel>(SpeciesDetailsLabelPath);
+        speciesPreview = GetNode<SpeciesPreview>(SpeciesPreviewPath);
+        hexesPreview = GetNode<CellHexesPreview>(HexPreviewPath);
+
+        UpdateCurrentWorldDetails();
+    }
+
+    public override void UpdateCurrentWorldDetails()
+    {
+        if (CurrentGame == null)
+            return;
+
+        RebuildTree();
+    }
+
+    public override void OnNavigationPanelSizeChanged(bool collapsed)
+    {
+    }
+
+    /// <summary>
+    ///   Clear and then rebuild the evolutionary tree each time we open the page. This way, we ensure the tree is
+    ///   always up to date.
+    /// </summary>
+    public void RebuildTree()
+    {
+        if (!Visible)
+            return;
+
+        // TODO fix the tree for freebuild?
+        if (CurrentGame!.FreeBuild)
+        {
+            evolutionaryTree.Visible = false;
+            disabledInFreebuild.Visible = true;
+            return;
+        }
+
+        // Building the tree relies on the existence of a full history of generations stored in the current game. Since
+        // we only started adding these in 0.6.0, it's impossible to build a tree in older saves.
+        // TODO avoid an ugly try/catch block by actually checking the original save version?
+        try
+        {
+            evolutionaryTree.Clear();
+            speciesHistoryList.Clear();
+
+            evolutionaryTree.Init(CurrentGame.GameWorld.PlayerSpecies);
+            InitFirstGeneration();
+
+            // A possible next step would be to rebuild only when the Thriveopedia as a whole is opened.
+            foreach (var generation in CurrentGame.GameWorld.GenerationHistory)
+            {
+                var record = generation.Value;
+                evolutionaryTree.UpdateEvolutionaryTreeWithRunResults(
+                    record.AutoEvoResult, record.Generation, record.TimeElapsed);
+                speciesHistoryList.Add(record.AllSpecies);
+            }
+        }
+        catch (KeyNotFoundException)
+        {
+            evolutionaryTree.Visible = false;
```

Comment:
Or at least print a TODO error here about the problem. And I haven't looked at the text but I hope the disabledInFreebuild error message also says that it may be because of an older save?

Reply:
Yes, maybe that's the wrong name as it covers both scenarios. I'll update that.